File: __iOS_BACKUP__/ShapeByte/shared/progress/data/Progress.swift
//
//  Progress.swift
//  ShapeByte
//
//  Created by Lang, Stefan [Shape Byte Tech] on 28.07.24.
//

import Foundation

struct Progress: Equatable {
    static let zero = Progress(0)
    static let complete = Progress(1)

    let value: CGFloat // 0..1

    // TODO: consider method with rounding rule
    var absoluteValue: Int { // 0 .. 100
        Int((value * 100.0).rounded())
    }

    init(_ value: CGFloat) {
        self.value = Self.adjustProgress(value)
    }

    // TODO: consider rounding rule
    init(absoluteValue: Int) {
        let relativeValue = CGFloat(absoluteValue) / 100.0
        self.init(relativeValue)
    }

    private static func adjustProgress(_ progress: CGFloat) -> CGFloat {
        let retVal: CGFloat = min(1.0, progress)
        return retVal
    }
}

File: shared/src/commonMain/kotlin/de/stefan/lang/shapebyte/utils/Progress.kt
package de.stefan.lang.shapebyte.utils
import kotlin.math.roundToInt

data class Progress(private val progress: Float) {
    companion object {
        val ZERO = Progress(0f)
        val COMPLETE = Progress(1f)

        const val ABSOLUTE: Int = 100

        // TODO: consider rounding rule
        fun with(current: Int, total: Int): Progress {
            val progress = current.toFloat() / total.toFloat()
            return Progress(progress)
        }

        // TODO: consider rounding rule
        fun with(current: UInt, total: UInt): Progress {
            val progress = current.toFloat() / total.toFloat()
            return Progress(progress)
        }
    }

    val value: Float = adjustProgress(progress)
    val absoluteValue: Int by lazy { (value * ABSOLUTE).roundToInt() }

    constructor(value: Int) : this(value.toFloat() / ABSOLUTE)

    private fun adjustProgress(value: Float): Float {
        return when {
            value < 0 -> 0f
            value > 1 -> 1f
            else -> value
        }
    }
}
